src/borg/testsuite/__init__.py

```diff
@@ -235,26 +235,63 @@ def _assert_dirs_equal_cmp(self, diff, ignore_bsdflags=False, ignore_xattrs=Fals
             self._assert_dirs_equal_cmp(sub_diff, ignore_bsdflags=ignore_bsdflags, ignore_xattrs=ignore_xattrs, ignore_ns=ignore_ns)
 
     @contextmanager
-    def fuse_mount(self, location, mountpoint, *options):
-        os.mkdir(mountpoint)
-        args = ['mount', location, mountpoint] + list(options)
-        self.cmd(*args, fork=True)
+    def fuse_mount(self, location, mountpoint=None, *options, **kwargs):
+        if mountpoint is None:
+            mountpoint = tempfile.mkdtemp()
+        else:
+            os.mkdir(mountpoint)
+        if 'fork' not in kwargs:
+            # For a successful mount, `fork = True` is required. The
+            # option to run the command with `fork = False` is still
+            # necessary to test mount failures like attempting to mount
+            # a read-only repo.
+            kwargs['fork'] = True
+        self.cmd('mount', location, mountpoint, *options, **kwargs)
         self.wait_for_mount(mountpoint)
         yield
         umount(mountpoint)
+        self.wait_for_mount(mountpoint, umount=True)
         os.rmdir(mountpoint)
         # Give the daemon some time to exit
-        time.sleep(.2)
+        time.sleep(0.2)
+
+    def wait_for_mount(self, mountpoint, timeout=5, umount=False):
+        """Wait until a path meets specified mount point status
 
-    def wait_for_mount(self, path, timeout=5):
-        """Wait until a filesystem is mounted on `path`
+        By default, checks if `mountpoint` is a mount point. To check
+        if `mountpoint` is *not* a mount point, set `umount = True`.
         """
         timeout += time.time()
         while timeout > time.time():
-            if os.path.ismount(path):
+            if os.path.ismount(mountpoint) != umount:
                 return
-            time.sleep(.1)
-        raise Exception('wait_for_mount(%s) timeout' % path)
+            time.sleep(0.1)
+        message = 'Waiting for %s of %s' % ('umount' if umount else 'mount', mountpoint)
+        raise TimeoutError(message)
+
+    @contextmanager
+    def read_only(self, path):
+        """Some paths need to be made read-only for testing
+
+        Using chmod to remove write permissions is not enough due to
+        the tests running with root privileges. Instead, the folder is
+        rendered immutable with chattr or chflags, respectively.
+        """
+        if sys.platform.startswith('linux'):
+            cmd_immutable = 'chattr +i "%s"' % path
+            cmd_mutable = 'chattr -i "%s"' % path
+        elif sys.platform.startswith('darwin'):
+            cmd_immutable = 'chflags schg "%s"' % path
+            cmd_mutable = 'chflags noschg "%s"' % path
+        else:
+            message = 'Testing read-only repos is currently not supported on your platform'
+            raise NotImplementedError(message)
```

Comment:
i will run those tests on misc. BSDs and openindiana, so does that mean my tests fill fail?

they rather should be skipped (or xfailed?) if we can not support this way of testing.

Reply:
Surely, those platforms have a similar command? Wouldn't that make more sense? We'll just need their [`sys.platform`](https://docs.python.org/3/library/sys.html#sys.platform) values.

Reply:
ok, i'll try to do some research for the other platforms.

Reply:
Like I commented above, `chflags` is a BSD command so that should not only work for OS X but for FreeBSD etc., too. However, for OpenIndiana/Solaris, I don't know. Is this the first time that there is a platform-depending issue? Instead of `NotImplementedError`, we could do `pytest.skip`, of course, as a quick fix, if you don't want to include that test for some platforms.

Reply:
openindiana:

sys.platform == 'sunos5'

chmod S+vimmutable file / chmod S-vimmutable file

Reply:
openbsd:

sys.platform == 'openbsd6'

`chflags schg` is a one-way ticket and only for root, so we can't use that.
`chflags (no)uchg` works.

Reply:
netbsd:

sys.platform == 'netbsd7'

schg: similar to openbsd, except for the one-way ticket.
(no)uchg works.

Reply:
freebsd:

sys.platform == 'freebsd10'

same as netbsd.

Reply:
for all other platforms, skip the test.

Reply:
:+1:

Reply:
https://github.com/borgbackup/borg/pull/5048/commits/5b52bcd0ae3e5ab067842d4600d1d164e97d1a1d Checks are passing, too